src/libponyrt/ds/hash.c

```diff
@@ -3,73 +3,135 @@
+  size_t mask)
 {
-  return ((uintptr_t)entry) > ((uintptr_t)DELETED);
+  return (current + map->size - (hash & mask)) & mask;
 }
 
-static void* search(hashmap_t* map, size_t* pos, void* key, hash_fn hash,
-  cmp_fn cmp)
+static void* search(hashmap_t* map, size_t* pos, void* key, size_t hash,
+  cmp_fn cmp, size_t* probe_length, size_t* oi_probe_length)
 {
-  size_t index_del = map->size;
-  size_t mask = index_del - 1;
+  size_t mask = map->size - 1;
 
-  size_t h = hash(key);
-  size_t index = h & mask;
+  size_t p_length = *probe_length;
+  size_t oi_p_length = 0;
+  size_t index = ((hash & mask) + p_length) & mask;
   void* elem;
+  size_t elem_hash;
 
   for(size_t i = 0; i <= mask; i++)
   {
-    elem = map->buckets[index];
+    elem = map->buckets[index].ptr;
+    elem_hash = map->buckets[index].hash;
 
     if(elem == NULL)
     {
-      if(index_del <= mask)
-        *pos = index_del;
-      else
-        *pos = index;
-
+      // empty bucket found
+      *pos = index;
+      *probe_length = p_length;
       return NULL;
-    } else if(elem == DELETED) {
-      /* some element was here, remember the first deleted slot */
-      if(index_del > mask)
-        index_del = index;
-    } else if(cmp(key, elem)) {
+    } else if(p_length >
+        (oi_p_length = get_probe_length(map, elem_hash, index, mask))) {
+      // our probe length is greater than the elements probe length
+      // we would normally have swapped so return this position
       *pos = index;
+      *probe_length = p_length;
+      *oi_probe_length = oi_p_length;
+      return NULL;
+    } else if((hash == elem_hash) && cmp(key, elem)) {
+      // element found
+      *pos = index;
+      *probe_length = p_length;
       return elem;
     }
 
-    index = (h + ((i + (i * i)) >> 1)) & mask;
+    index = (index + 1) & mask;
+    p_length++;
   }
 
-  *pos = index_del;
+  *pos = map->size;
+  *probe_length = p_length;
   return NULL;
 }
 
-static void resize(hashmap_t* map, hash_fn hash, cmp_fn cmp, alloc_fn alloc,
+static void resize(hashmap_t* map, cmp_fn cmp, alloc_fn alloc,
   free_size_fn fr)
 {
   size_t s = map->size;
-  void** b = map->buckets;
+  size_t c = map->count;
+  hashmap_entry_t* b = map->buckets;
+  bitmap_t* old_item_bitmap = map->item_bitmap;
   void* curr = NULL;
 
   map->count = 0;
-  map->size = (s < 8) ? 8 : s << 3;
-  map->buckets = (void**)alloc(map->size * sizeof(void*));
-  memset(map->buckets, 0, map->size * sizeof(void*));
+  map->size = (s < MIN_HASHMAP_SIZE) ? MIN_HASHMAP_SIZE : s << 3;
+
+  // use a single memory allocation to exploit spatial memory/cache locality
+  size_t bitmap_size = map->size/HASHMAP_BITMAP_TYPE_SIZE +
+    (map->size%HASHMAP_BITMAP_TYPE_SIZE==0?0:1);
+  void* mem_alloc = alloc((bitmap_size * sizeof(bitmap_t)) +
+    (map->size * sizeof(hashmap_entry_t)));
+  memset(mem_alloc, 0, (bitmap_size * sizeof(bitmap_t)) +
+   (map->size * sizeof(hashmap_entry_t)));
```

Comment:
Do the elements need to be NULL'd out, or would the bitmap being zero'd be enough?